torch/lib/c10d/TCPStore.cpp

```diff
@@ -522,6 +727,27 @@ bool TCPStore::deleteKey(const std::string& key) {
   return (numDeleted == 1);
 }
 
+void TCPStore::watchKey(
+    const std::string& key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)>
+        callback) {
+  std::string regKey = regularPrefix_ + key;
+
+  // Register callback with TCPStoreMasterDaemon to call TCPStoreWorkerDaemon on
+  // key change
+  tcpStoreWorkerDaemon_->addCallback(regKey, callback);
+  tcputil::sendValue<QueryType>(listenSocket_, QueryType::WATCH_KEY);
+  tcputil::sendString(listenSocket_, regKey);
```

Comment:
Sending the request is not protected by a lock. Does this mean it is possible to send multiple register callback requests concurrently? If that's true, it also means that the following critical section can match with a different request? E.g., say we have two registering requests, and the response of the 2nd request comes back first and unblocks the first request's CV. Does this mean when `watchKey` returns, it does not always guarantee that the registration is done on the master?

Reply:
Updated the TCPStore::watchKey method to use a lock so only one register callback request can be sent per process at a time.